### PR TITLE
File upload progress

### LINKED ISSUE #648: Progress callback for S3Object
Hi, I would like to indicate progress of a large file upload to my users. Currently it is not supported with the SDK. If I wanted to use a single-part upload only (I don't) I could use the block form of `S3Object#write` and report how many bytes I've fed to the SDK, but that's not an actual upload - at that point the data is still waiting to be sent. For multipart upload there's not even that.

What do you think about adding a callback to that method? Something like:

```
progress = -> bytes { ... }
object.write file: path, progress_callback: progress
```

### PR DESCRIPTION
**Pull request by alextwoods**
Fixes #648 

Example of current usage:
```ruby
object = S3::Resource.new.bucket(bucket_name).object(object_name)
progress = Proc.new { |bytes, totals| puts bytes.map.with_index { |b, i| "Part #{i+1}: #{b} / #{totals[i]}"}.join(' ') + " Total: #{100.0 * bytes.sum / totals.sum }%" }
object.upload_file(large_file, progress_callback: progress)
```

Outputs:
```
Part 1: 0 / 5242880 Part 2: 32768 / 5242880 Part 3: 0 / 5242880 Part 4: 147456 / 1249426  Total: 1.061510775137757%
.....
Part 1: 1245184 / 5242880 Part 2: 1196032 / 5242880 Part 3: 1540096 / 5242880 Part 4: 1163264 / 1249426 Total: 30.30130758120507%
.....
Part 1: 5242880 / 5242880 Part 2: 5242880 / 5242880 Part 3: 5242880 / 5242880 Part 4: 1249426 / 1249426 Total: 100.0%

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

1. To make sure we include your contribution in the release notes, please make sure to add description entry for your changes in the "unreleased changes" section of the `CHANGELOG.md` file (at corresponding gem). For the description entry, please make sure it lives in one line and starts with `Feature` or `Issue` in the correct format.

2. For generated code changes, please checkout below instructions first:
  https://github.com/aws/aws-sdk-ruby/blob/master/CONTRIBUTING.md

Thank you for your contribution!

### CODE REVIEW

File: gems/aws-sdk-s3/lib/aws-sdk-s3/multipart_file_uploader.rb [hunk 1]
```diff
@@ -187,6 +196,23 @@ def to_a
...
+          @total_sizes = parts.map { |p| p[:body].size }
+          @progress_callback = progress_callback
+        end
+
+        def call(part_number, bytes_read)
+          # part numbers start at 1
+          @bytes_sent[part_number - 1] = bytes_read
+          @progress_callback.call(@bytes_sent, @total_sizes)
+        end
+
```
Comment (mullermp) [inline gems/aws-sdk-s3/lib/aws-sdk-s3/multipart_file_uploader.rb:214]: space

File: gems/aws-sdk-s3/lib/aws-sdk-s3/multipart_file_uploader.rb [hunk 2]
```diff
@@ -187,6 +196,23 @@ def to_a
         end
 
       end
+
+      # @api private
+      class MultipartProgress
+        def initialize(parts, progress_callback)
+          parts = parts.to_a
```
Comment (mullermp) [inline gems/aws-sdk-s3/lib/aws-sdk-s3/multipart_file_uploader.rb:203]: Check to see if this is done anywhere else, and if so, it maybe makes sense to implement size and map.

File: gems/aws-sdk-s3/lib/aws-sdk-s3/customizations/object.rb [hunk 3]
```diff
@@ -312,6 +312,8 @@ def upload_stream(options = {}, &block)
       #   multipart uploads. This option is not used if the file is smaller than
       #   `:multipart_threshold`.
       #
+      # @option options [Proc] :progress_callback - TBD
```
Comment (mullermp) [inline gems/aws-sdk-s3/lib/aws-sdk-s3/customizations/object.rb:315]: progress_callback makes sense.. but it's also limiting to what can be done here. would this be only used in that case? if not then probably a new generic name would be better

File: gems/aws-sdk-s3/lib/aws-sdk-s3/file_uploader.rb [hunk 4]
```diff
@@ -49,11 +51,19 @@ def open_file(source)
       end
 
       def put_object(source, options)
+        if options.include? :progress_callback
+          options[:on_chunk_sent] =
+            single_part_progress(options.delete(:progress_callback))
+        end
```
Comment (mullermp) [inline gems/aws-sdk-s3/lib/aws-sdk-s3/file_uploader.rb:57]: can just do `if (callback = options.delete(:progress_callback))`

File: gems/aws-sdk-s3/lib/aws-sdk-s3/file_uploader.rb [hunk 5]
```diff
@@ -49,11 +51,19 @@ def open_file(source)
...
         open_file(source) do |file|
           @client.put_object(options.merge(body: file))
         end
       end
 
+      def single_part_progress(progress_callback)
+        Proc.new { |_chunk, bytes_read, total_size| progress_callback.call([bytes_read], [total_size]) }
```
Comment (mullermp) [inline gems/aws-sdk-s3/lib/aws-sdk-s3/file_uploader.rb:64]: multiline (and remove space at end)

File: gems/aws-sdk-s3/lib/aws-sdk-s3/multipart_file_uploader.rb [hunk 6]
```diff
@@ -66,9 +67,10 @@ def complete_upload(upload_id, parts, options)
       end
 
       def upload_parts(upload_id, source, options)
+        puts "options: #{options}"
```
Comment (mullermp) [inline gems/aws-sdk-s3/lib/aws-sdk-s3/multipart_file_uploader.rb:70]: puts

File: gems/aws-sdk-s3/lib/aws-sdk-s3/multipart_file_uploader.rb [hunk 7]
```diff
@@ -127,13 +129,20 @@ def upload_part_opts(options)
...
+        if options.include? :progress_callback
+          progress = MultipartProgress.new(pending, options[:progress_callback])
+        end
         @thread_count.times do
           thread = Thread.new do
             begin
               while part = pending.shift
-                resp = @client.upload_part(part)
+                if progress
+                  upload_part_options[:on_chunk_sent] = Proc.new { |_chunk, bytes, _total| progress.call(part[:part_number], bytes) }
```
Comment (mullermp) [inline gems/aws-sdk-s3/lib/aws-sdk-s3/multipart_file_uploader.rb:143]: multiline - could also maybe clean up similarly to the put_object case?

File: gems/aws-sdk-s3/lib/aws-sdk-s3/multipart_file_uploader.rb (hunk 2 above)
Comment (alextwoods) [reply gems/aws-sdk-s3/lib/aws-sdk-s3/multipart_file_uploader.rb:203]: `PartList#to_a` is also used when returning the completed parts ordered by `:part_number` which wouldn't be solved by adding `map` - instead, I'm adding a specific method for `part_sizes` as I don't think we want to expose map and I do prefer not using the `to_a` method here.

File: gems/aws-sdk-s3/CHANGELOG.md [hunk 8]
```diff
@@ -1,6 +1,8 @@
 Unreleased Changes
 ------------------
 
+* Feature - Add progress_callback to Object#upload to support reporting of upload progress. (#648)
```
Comment (mullermp) [inline gems/aws-sdk-s3/CHANGELOG.md:4]: backticks

File: gems/aws-sdk-s3/spec/object/upload_file_spec.rb [hunk 9]
```diff
@@ -123,6 +139,21 @@ module S3
...
+              double(etag: 'etag')
+            end
+            callback = proc do |bytes, totals|
+              expect(bytes.size).to eq(4)
+              expect(totals.size).to eq(4)
+            end
+            object.upload_file(seventeen_meg_file, content_type: 'text/plain', progress_callback: callback)
+          end
+
+
```
Comment (mullermp) [inline gems/aws-sdk-s3/spec/object/upload_file_spec.rb:156]: extra space

**Comment by jethrodaniel** [comment]
This is great - when will it be released?

**Comment by alextwoods** [comment]
It was released in [1.75.0](https://github.com/aws/aws-sdk-ruby/blob/master/gems/aws-sdk-s3/CHANGELOG.md#1750-2020-07-21) of aws-sdk-s3 (on 2020-07-21).

**Pull request merged**